### DISCUSSION TITLE
Hydration mismatch—state mismatch?

### DISCUSSION THREAD

**Discussion by henningko** [nuxt/nuxt#17015] [Questions] [answered]
Heya!

I'm setting up an app that uses supabase for auth, but that also uses Nuxt's server to make some requests to Supabase (e.g. properly formatting responses that shouldn't happen client-side).

However, I'm seeing hydration issues, which probably come from a state mismatch:

```
[Vue warn]: Hydration node mismatch:
- Client vnode: div 
- Server rendered DOM: 
<form class="row flex flex-center"> <empty string> 
  at <Index onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > > 
  at <BaseTransition mode="out-in" appear=false persisted=false  ... > 
  at <Transition name="page" mode="out-in" > 
  at <RouterView > 
  at <NuxtPage> 
  at <App> 
  at <NuxtRoot>
```

**Edit:** One thing that's worth mentioning is that this only occurs when the page is refreshed by an already logged-in user. Without an already logged in user, the user sees the auth form, and after signing in, sees the content, all without a hydration mismatch.

Which probably stems from a problem with my if/else, as the form referenced up top is the form shown if `isAuthenticated()` fails:
```vue
<template>
  <div class="w-full min-h-screen bg-white dark:bg-black dark:text-white p-8">
    <div v-if="isAuthenticated()">
      <Logout />
      <h1 class="text-7xl font-bold mb-8">Brief.</h1>
      <Content />
    </div>
    <Auth v-else />
  </div>
</template>
<script setup lang="ts">
const { $supabase } = useNuxtApp();
const loading = ref(false);
const { isAuthenticated } = await useAuth();
</script>
```

`useAuth.ts`
```ts
import type { Session } from "@supabase/supabase-js";

// session is null when not logged in
const userSession = ref<Session | null>(null);
const userCookieSet = ref<boolean | null>(null);

export default async () => {
  const isAuthenticated = () => {
    return userSession.value?.user && userCookieSet.value;
  };

  return {
    userSession,
    userCookieSet,
    isAuthenticated,
  };
};
```

whereby `userSession` is set by `supabase.auth`, and a subsequent request to `/api/auth` ensures that a cookie is being set so that Nuxt's API can make requests to supabase on behalf of the user.

Full repo: https://github.com/henningko/briefxyz/

New to Nuxt3/SSR/Supabase and biting off more than I can chew, probably, so please don't hold back with criticism :-)

**Environment**
<details>
<summary>macOS/Nuxt 3.0.0/Node 17.4 (click for more)</summary>

* Operating System: `macOS`
* Node Version:     `v17.4.0`
* Nuxt Version:     `3.0.0-27398533.8edd481`
* Package Manager:  `Yarn`
* Bundler:          `Vite`
* User Config:      `buildModules`, `RuntimeConfig`
* Runtime Modules:  `-`
* Build Modules:    `nuxt-windicss@2.2.2`
</details>

**Comment by sewalsh** [comment]
Does wrapping `Auth` in a `div` with the `v-else` silence the warning?

```js
  <div v-if="isAuthenticated()">
    <Logout />
    <h1 class="text-7xl font-bold mb-8">Brief.</h1>
    <Content />
  </div>
  <div v-else>
    <Auth />
  </div>
```

**Reply by henningko** [reply]
Unfortunately, that creates two mismatch warnings, at the node and the child:
```
[Vue warn]: Hydration node mismatch:
- Client vnode: div 
- Server rendered DOM: 
<form class="row flex flex-center">
 <empty string> 
  at <Logout> 
  at <Index onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > > 
  at <BaseTransition mode="out-in" appear=false persisted=false  ... > 
  at <Transition name="page" mode="out-in" > 
  at <RouterView > 
  at <NuxtPage> 
  at <App> 
  at <NuxtRoot> [runtime-core.esm-bundler.js:38:16](http://localhost:3000/_nuxt/node_modules/@vue/runtime-core/dist/runtime-core.esm-bundler.js)
[Vue warn]: Hydration children mismatch in <div>: server rendered element contains fewer child nodes than client vdom. 
  at <Index onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > > 
  at <BaseTransition mode="out-in" appear=false persisted=false  ... > 
  at <Transition name="page" mode="out-in" > 
  at <RouterView > 
  at <NuxtPage> 
  at <App> 
  at <NuxtRoot>
```

One thing that's worth mentioning is that this only occurs when refreshing when the user is already logged in. Without an already logged in user, the user sees the auth form, and after signing in, sees the content, all without a hydration mismatch.

**Comment by wobsoriano** [comment]
What if you update your composable to look like this?

```ts
import type { Session } from "@supabase/supabase-js";

export default () => {
  const userSession = ref<Session | null>(null);
  const userCookieSet = ref<boolean | null>(null);
  const isAuthenticated = ref(false)

  watch([userSession, userCookieSet], ([session, cookie]) => {
    if (session?.user && cookie) {
      isAuthenticated.value = true;
    } else {
      isAuthenticated.value = false
    }
  })

  return {
    userSession,
    userCookieSet,
    isAuthenticated,
  };
};
```

Then your template

```html
<script setup lang="ts">
const { isAuthenticated } = useAuth();
</script>

<template>
  <div class="w-full min-h-screen bg-white dark:bg-black dark:text-white p-8">
    <div v-if="isAuthenticated">
      <Logout />
      <h1 class="text-7xl font-bold mb-8">Brief.</h1>
      <Content />
    </div>
    <Auth v-else />
  </div>
</template>
```

**Reply by henningko** [reply]
Unfortunately, this won't show the content at all anymore. Eventually, it seems, everything evaluates fine:
<img width="695" alt="image" src="https://user-images.githubusercontent.com/5890910/152698066-43c14c63-8bb5-408a-8204-35af6129faa4.png">

... but it keeps showing the form; and logging `isAuthenticated` prints `false` on both the server and the client.

```vue
<script setup lang="ts">
const { isAuthenticated } = useAuth();
console.log(isAuthenticated.value);
</script>
<template>
  <div class="w-full min-h-screen bg-white dark:bg-black dark:text-white p-8">
    <div v-if="isAuthenticated">
      <Logout />
      <h1 class="text-7xl font-bold mb-8">Brief.</h1>
      <Content />
    </div>
    <div v-else>
      <Auth />
    </div>
  </div>
</template>
```

**Reply by wobsoriano** [reply]
try to watch the value of `isAuthenticated` because it will always be false in your sample code

```html
<script setup lang="ts">
const { isAuthenticated } = useAuth();
console.log(isAuthenticated.value);

watch(isAuthenticated, (val) => {
  console.log(val)
})
</script>
```

**Reply by henningko** [reply]
Hmm, interestingly, the watcher never fires, even though `isAuthenticated` certainly changes.

When adding this to `app.vue` (and not `pages/index.vue`), it reliably reacts to changes and logs the correct values.
```js
watch(isAuthenticated, (val) => {
  console.log(val)
})
```

Which leads me to believe that there's something about sticking composables into template files that impedes reactivity?

**Reply by henningko** [reply]
If I move everything into `app.vue`, the content reappears, but there's still a hydration warning, and a new error: `An error occurred in hook walkComponentTree [backend.js:710:19]`

Also, the `console.log` statements here will always produce "false" on the backend, and correctly change on the frontend—which likely explains the hydration problem?

```vue
<template>
  <div class="w-full min-h-screen bg-white dark:bg-black dark:text-white p-8">
    <div v-if="isAuthenticated">
      <Logout />
      <h1 class="text-7xl font-bold mb-8">Brief.</h1>
      <Content />
    </div>
    <Auth v-else />
  </div>
</template>
<script setup lang="ts">
import "./assets/css/global.css";
const { $supabase } = useNuxtApp();
const { userSession, userCookieSet, isAuthenticated } = useAuth();
console.log(isAuthenticated.value);
watch(isAuthenticated, (val) => {
  console.log(val);
});
// Initial setup
userSession.value = await $supabase.auth.session();
if (userSession.value) {
  ({ auth: userCookieSet.value } = await $fetch("api/auth", {
    method: "POST",
    body: { session: userSession.value, event: "TOKEN_REFRESHED" },
  }));
}
$supabase.auth.onAuthStateChange(async (event, session) => {
  userSession.value = session;
  ({ auth: userCookieSet.value } = await $fetch("api/auth", {
    method: "POST",
    body: { event, session },
  }));
});
</script>
```


Edit: Alternatively, adapting `useAuth.ts`based on your suggestion to the following (moving definitions outside of the export function) works with having the logic in `pages/index.vue`—e.g. content shows up—but doesn't remedy the hydration warning 🤯
```js
import type { Session } from "@supabase/supabase-js";
const userSession = ref<Session | null>(null);
const userCookieSet = ref<boolean | null>(null);
const isAuthenticated = ref(false);

watch([userSession, userCookieSet], ([session, cookie]) => {
  if (session?.user && cookie) {
    isAuthenticated.value = true;
  } else {
    isAuthenticated.value = false;
  }
});

export default () => {
  return {
    userSession,
    userCookieSet,
    isAuthenticated,
  };
};
```

**Comment by jdalrymple** [comment]
@henningko Did you manage to sort this out?

**Reply by henningko** [reply]
[Nuxt Supabase](https://supabase.nuxtjs.org) was released around the time Nuxt 3 RC 1 came around, and following the steps outlined [here](https://supabase.nuxtjs.org/usage/composables/use-supabase-user) to protect routes took care of the problem.

I also moved the API routes created in `/server/` into Supabase Edge Functions.

So, I never was quite able to fix the problem, but development of other tools caught up with me :-)

**Comment by henningko** [comment, marked as answer]
Update: Using [Nuxt Supabase](https://supabase.nuxtjs.org) and moving endpoints from `/server` to Supabase Edge Functions to care of my issues.
